File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        parentheses_map = {'}': '{', ']': '[', ')': '('}
        stack = []
        
        for parentheses in s:
            if parentheses in parentheses_map:
                if not stack or stack[-1] != parentheses_map[parentheses]:
                    return False
                stack.pop()
            else:
                stack.append(parentheses)
        return stack == []

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    private static final HashMap<Character, Character> MAPPINGS;
    
    static {
        MAPPINGS = new HashMap<>();
        MAPPINGS.put(']', '[');
        MAPPINGS.put('}', '{');
        MAPPINGS.put(')', '(');
    }
    
    
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        
        for (char ch : s.toCharArray()) {
            
            if (MAPPINGS.containsKey(ch)) {
                char topElement = stack.isEmpty() ? '#' : stack.pop();

                if (topElement != MAPPINGS.get(ch))
                    return false;
            } else {
                stack.push(ch);
            }
            
        } 
        
        return stack.isEmpty();
        
        
    }
}